docs/proposals/dramatically-simplify-cluster-creation.md

```diff
@@ -0,0 +1,222 @@
+
+3. *Add-ons*: Now that basic cluster functionality is working, installing add-ons such as DNS or a pod network (should be possible using kubectl apply).
+
+Notably, this project is *only* working on dramatically improving 2 and 3 from the perspective of users typing commands directly into root shells of servers. The reason for this is that there are a great many different ways of provisioning servers, and users will already have their own preferences.
+
+What's more, once we've radically improved the user experience of 2 and 3, it will make the job of tools that want to do all three much easier.
+
+## User stories
+
+### Phase I
+
+**_In time to be an alpha feature in Kubernetes 1.4._**
+
+* *Install*: As a potential Kubernetes user, I can deploy a Kubernetes 1.4 cluster on a handful of computers running Linux and Docker by typing two commands on each of those computers. The process is so simple that it becomes obvious to me how to easily automate it if I so wish.
+
+* *Pre-flight check*: If any of the computers don't have working dependencies installed (e.g. bad version of Docker, too-old Linux kernel), I am informed early on and given clear instructions on how to fix it so that I can keep trying until it works.
+
+* *Control*: Having provisioned a cluster, I can gain user credentials which allow me to remotely control it using kubectl.
+
+* *Install-addons*: I can select from a set of recommended add-ons to install directly after installing Kubernetes on my set of initial computers with kubectl apply.
+
+* *Add-node*: I can add another computer to the cluster.
+
+* *Secure*: As an attacker with (presumed) control of the network, I cannot add malicious nodes I control to the cluster created by the user. I also cannot remotely control the cluster.
+
+### Phase II
+
+**_In time for Kubernetes 1.5:_**
+*Everything from Phase I as beta/stable feature, everything else below as beta feature in Kubernetes 1.5.*
+
+* *Upgrade*: Later, when Kubernetes 1.4.1 or any newer release is published, I can upgrade to it by typing one other command on each computer.
+
+* *HA*: If one of the computers in the cluster fails, the cluster carries on working. I can find out how to replace the failed computer, including if the computer was one of the masters.
+
+## Top-down view: UX for Phase I items
+
+We will introduce a new binary, kubeadm, which ships with the Kubernetes OS packages (and binary tarballs, for OSes without package managers).
+
+```
+laptop$ kubeadm --help
+
+kubeadm: bootstrap a kubernetes cluster.
+
+Global options:
+
+    --discovery    Discovery mechanism to use. Choose from: gossip,
+                   out-of-band. Default: gossip.
+
+Subcommands:
+    init master    Run this on the first server you deploy onto.
+
+    join node      Run this on other servers to join an existing cluster.
+
+    user create    Create user and fetch authentication config for them.
+```
+
+### Install
+
+*On first machine:*
+
+```
+master# kubeadm init master
```

Comment:
> Will `kubeadm init` download anything, or will it assume that everything has been downloaded?

Great question. It will cause static pods to be written which will download other kubernetes components and run them as pods. If a user wants a completely isolated, disconnected install, they would need to pre-fetch these containers, or bake them into their AMIs or what-have-you. Note that working when disconnected from the Internet is one of the key benefits of the approach advocated for in #30361.